irc_send.c

```diff
@@ -293,15 +287,26 @@ void irc_send_who(irc_t *irc, GSList *l, const char *channel)
 	gboolean is_channel = strchr(CTYPES, channel[0]) != NULL;
 
 	while (l) {
-		irc_user_t *iu = l->data;
+		irc_user_t *iu;
+
+		/* Null terminated string with three chars, respectively:
+		 * { <H|G>, <@|%|+|\0>, \0 } */
+		char status_prefix[3] = {0};
```

Comment:
hmmm, from xep-0045:

>MUC also defines two different associations: long-lived affiliations and session-specific roles. These two association types are distinct from each other in MUC, since an affiliation lasts across visits, while a role lasts only for the duration of a visit. In addition, there is no one-to-one correspondence between roles and affiliations; for example, someone who is not affiliated with a room may be a (temporary) moderator, and a member may be a participant or a visitor in a moderated room.

Roles: Moderator, participant, visitor, none

Affiliations: Owner, admin, member, outcast, none

And affiliations define the default role when joining a channel (owner and admin become moderators, members are always participant, none depends on how open the channel is, outcast is always banned)

I'd say roles are like modes, and affiliations are like chanserv access levels. Moderator maps to @, the rest map to no mode.

Anyway the groupchat code doesn't even support this at all yet. And when mapping this part of IM services, i'd rather go for the simplest solution